cmake: emu: qemu: introduce `CONFIG_QEMU_GDBSERVER_LISTEN_DEV` so that gdbserver listeners other than `-s`  can be set

alpsayin opened the pull request:
First, we introduce a new string option to pass into QEMU invocation so that gdbserver listeners other than `-s` (tcp::1234) can be set while setting its default to `"tcp::1234"` for backwards compatibility.

Then we remove the enforced `-s` and set `-gdb` but only if above-mentioned Kconfig is NOT an empty string.
This allows users to utilize `QEMU_EXTRA_FLAGS` from shell environment. 
Otherwise, trying to inject a `-gdb` flag from shell causes a fatal clash that occurs between `-s` and `-gdb` during QEMU invocation.

Signed-off-by: Alp Sayin <alpsayin@gmail.com>

galak reviewed (changes requested):
Can you add this into the docs about qemu debug

alpsayin commented:
It just came to me that this is probably more useful as a Kconfig option. 
If anyone wants to inject at CMake level it'll allow that too.
I'll update the doc too accordingly.

alpsayin committed:
cmake: emu: qemu: start QEMU gdbserver only if relevant Kconfig is set

This removes the enforced `-s` option and only passes in a `-gdb` only if
CONFIG_QEMU_GDBSERVER_LISTEN_DEV is set. When unset, it allows users to
utilize QEMU_EXTRA_FLAGS from environment preventing a clash that occurs
if `-s` and `-gdb` are both passed to QEMU invocation.

Signed-off-by: Alp Sayin <alpsayin@gmail.com>

alpsayin committed:
doc: application development: brief about setting QEMU gdbserver listener

Briefly explain how a user can change their gdbserver listening device via
the Kconfig option or via shell environment. User is reminded about caveats
regarding using environment variables with CMake systems. Finally, another
invokation of QEMU-with-gdb via west from root directory is presented.

Signed-off-by: Alp Sayin <alpsayin@gmail.com>

stephanosio commented:
> Done. But I couldn't find anywhere how to "resolve" this conversation. Is this on you?

@alpsayin Yes, @galak either needs to approve or dismiss his review.

nordicjm commented on boards/Kconfig:
```diff
@@ -99,6 +99,18 @@ config QEMU_UEFI_BOOT
+	help
+	  This options is passed onto QEMU as a parameter to `-gdb` option.
+	  The default value is equivalent to `-s` which is a shorthand for
+		`-gdb tcp::1234`. An empty value omits the `-gdb` parameter altogether.
+		This allows the injection of `-gdb` parameter from other sources such
+		as the `QEMU_EXTRA_FLAGS` environment variable. Refer to application
+		development doc and/or QEMU invocation doc for more info.
```
This is not right, it was correct before, 1 tab followed by 2 spaces

alpsayin replied:
Wow, you're fast! Yes, I noticed it immediately too (burnt by my editor's consistent tabbing)

alpsayin committed:
boards: kconfig: qemu: introduce QEMU_GDBSERVER_LISTEN_DEV

Introduce new string option to pass into QEMU invocation so that gdbserver
listeners other than `-s` (tcp::1234) can be set.

Signed-off-by: Alp Sayin <alpsayin@gmail.com>

alpsayin commented:
🤦🏼 I'm sorry, I pressed a button and it removed @nashif @tejlmand @carlescufi and @jeremybettis. I blame git for this one.

nordicjm commented:
> 🤦🏼 I'm sorry, I pressed a button and it removed @nashif @tejlmand @carlescufi and @jeremybettis. I blame git for this one.

Github, full of bugs, never fixed - the microsoft way
